blocks/monitoring/aggregation.py

```diff
@@ -138,11 +138,62 @@ def get_aggregator(self):
         return aggregator
 
 
-def mean(numerator, denominator=1.):
+def mean(numerator, denominator=1., name=None):
     """Mean of quantity (numerator) over a number (denominator) values."""
     variable = numerator / denominator
     variable.tag.aggregation_scheme = Mean(numerator, denominator)
-    variable.name = numerator.name
+    if name is not None:
```

Comment:
You are correct about theano behavior, still I guess that since adding 
an explicit aggregation scheme is a strong sign of creating an 
end-product which should have a meaningful name, it is the readability 
of the code that uses these constructs that matters.

We can meet in the middle by delegating the name handling to an 
auxiliary function and calling it in all the aggregation schemes.

I find it much nicer to write in the end code:
```
extension.monitor(mean(smth, denom), name="cost-per-token"))
```

rather than:
```
per_token_cost = mean(smth, denom)
per_token_cost.name = "cost-per-token"
extension.monitor(per_token_cost)
```

and this is the code bloat I want to get rid of.

Dmitry Bogdanov wrote:
>
> In blocks/monitoring/aggregation.py
> <https://github.com/mila-udem/blocks/pull/780#discussion_r36044151>:
>
> > """Mean of quantity (numerator) over a number (denominator) values."""
> > variable = numerator / denominator
> > variable.tag.aggregation_scheme = Mean(numerator, denominator)
> > - variable.name = numerator.name
> > + if name is not None:
>
> |matrix| and |shared| are a bit different because they create input
> variables from scratch. This way it is natural for them to accept
> names. Correct me I am wrong, but no Theano expression (i.e. a
> function that takes input variables and creates output variables)
> supports |name| keyword, and if we want to mimic Theano here, we
> should not have it as well.
>
> I am quite disappointed in |named_copy|, to be honest, I think it only
> deteriorates readability.
>
> Also now we have |mean| and |sum|, but I also anticipate |min| and
> |max|, and this little piece of code that sets the name will have to
> be copied even more times. I think that the ratio between the
> convenience this feature brings and the code bloat it brings is not
> favorable.
>
> I would be maybe a bit less skeptical if the |name| argument was added
> by a decorator, reducing the need for copy-paste. This would however
> decrease the readability compared to straight-forward adding the name
> to the returned variable, because one would have to read the code of
> the decorator to see how |name| is handled.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/780/files#r36044151>.